src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -48,60 +49,96 @@ internal static SafeChannelBindingHandle QueryChannelBinding(SafeSslHandle conte
+                    // create the handle, it's rooted by the using, no one else has a reference to it, etc.
+                    Ssl.SetProtocolOptions(innerContext.DangerousGetHandle(), protocols);
 
-                if (remoteCertRequired)
-                {
-                    Debug.Assert(isServer, "isServer flag should be true");
-                    Ssl.SslCtxSetVerify(innerContext,
-                        s_verifyClientCertificate);
+                    // The logic in SafeSslHandle.Disconnect is simple because we are doing a quiet
+                    // shutdown (we aren't negotiating for session close to enable later session
+                    // restoration).
+                    //
+                    // If you find yourself wanting to remove this line to enable bidirectional
+                    // close-notify, you'll probably need to rewrite SafeSslHandle.Disconnect().
+                    // https://www.openssl.org/docs/manmaster/ssl/SSL_shutdown.html
+                    Ssl.SslCtxSetQuietShutdown(innerContext);
 
-                    //update the client CA list 
-                    UpdateCAListFromRootStore(innerContext);
-                }
+                    if (!Ssl.SetEncryptionPolicy(innerContext, policy))
+                    {
+                        throw new PlatformNotSupportedException(SR.Format(SR.net_ssl_encryptionpolicy_notsupported, policy));
+                    }
 
-                context = SafeSslHandle.Create(innerContext, isServer);
-                Debug.Assert(context != null, "Expected non-null return value from SafeSslHandle.Create");
-                if (context.IsInvalid)
-                {
-                    context.Dispose();
-                    throw CreateSslException(SR.net_allocate_ssl_context_failed);
+                    if (hasCertificateAndKey)
+                    {
+                        SetSslCertificate(innerContext, certHandle, certKeyHandle);
+                    }
+
+                    if (remoteCertRequired)
+                    {
+                        Debug.Assert(isServer, "isServer flag should be true");
+                        Ssl.SslCtxSetVerify(innerContext,
+                            s_verifyClientCertificate);
+
+                        //update the client CA list 
+                        UpdateCAListFromRootStore(innerContext);
+                    }
+
+                    context = SafeSslHandle.Create(innerContext, isServer);
+                    Debug.Assert(context != null, "Expected non-null return value from SafeSslHandle.Create");
+                    if (context.IsInvalid)
+                    {
+                        context.Dispose();
+                        throw CreateSslException(SR.net_allocate_ssl_context_failed);
+                    }
+
+                    if (hasCertificateAndKey)
+                    {
+                        using (X509Certificate2 cert = new X509Certificate2(certHandle.DangerousGetHandle()))
```

Comment:
This is really the only line that needs to be protected by the DangerousAddRef.

* Keeping the AddRef near the GetHandle shows why the AddRef is there.  As it stands right now, this could be refactored and no one would realize that it was the reason that the AddRef existed.
* When a SafeHandle is passed through a P/Invoke the AddRef/Release methods are automatically called (with the exceptions of an invalid/corrupt state accordingly).  Moving the call in tighter just makes this look closer to what the P/Invoke stub would be if we were calling a native method.
* The private key doesn't need a manual AddRef at all.